src/main/java/it/eng/idsa/businesslogic/processor/common/ContractAgreementProcessor.java

```diff
@@ -0,0 +1,56 @@
+package it.eng.idsa.businesslogic.processor.common;
+
+import org.apache.camel.Exchange;
+import org.apache.camel.Processor;
+import org.apache.commons.lang3.StringUtils;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.springframework.beans.factory.annotation.Value;
+import org.springframework.stereotype.Component;
+
+import de.fraunhofer.iais.eis.ContractAgreementMessage;
+import it.eng.idsa.businesslogic.processor.exception.ExceptionForProcessor;
+import it.eng.idsa.businesslogic.service.CommunicationService;
+import it.eng.idsa.multipart.domain.MultipartMessage;
+
+/**
+ * Processor that handles ContractAgreementMessage and sends ContractAgreement (payload of MultipartMessage) to UsageControl DataApp</br>
+ * This processor should be called before calling ClearingHouse registration processor, in order to be compliant with IDS flow.
+ * @author igor.balog
+ *
+ */
+@Component
+public class ContractAgreementProcessor implements Processor {
+	
+	private static final Logger logger = LogManager.getLogger(ContractAgreementProcessor.class);
+	private CommunicationService communicationService;
+	private String usageControlDataAppURL;
+	private String addPolicyEndpoint;
+	
+	public ContractAgreementProcessor(CommunicationService communicationService, 
+			@Value("${spring.ids.ucapp.baseUrl}") String usageControlDataAppURL,
+			@Value("${spring.ids.ucapp.addPolicyEndpoint}") String addPolicyEndpoint) {
+		this.communicationService = communicationService;
+		this.usageControlDataAppURL = usageControlDataAppURL;
+		this.addPolicyEndpoint = addPolicyEndpoint;
+	}
+
+	@Override
+	public void process(Exchange exchange) throws Exception {
+		MultipartMessage multipartMessage = exchange.getMessage().getBody(MultipartMessage.class);
+		
+		if(multipartMessage.getHeaderContent() instanceof ContractAgreementMessage) {
+			if(StringUtils.isNotBlank(multipartMessage.getPayloadContent())) {
+				String ucDataAppAddPolicyEndpoint = usageControlDataAppURL + addPolicyEndpoint;
+				logger.info("ContractAgreementMessage detected, sending payload to Usage Contol DataApp at '{}'", ucDataAppAddPolicyEndpoint);
+				String response = communicationService.sendDataAsJson(ucDataAppAddPolicyEndpoint, multipartMessage.getPayloadContent());
+				logger.info("UsageControl DataApp response {}", response);
+			} else {
+				logger.warn("Payload not present but mandatory for this logic");
+				throw new ExceptionForProcessor("Payload not present but mandatory for this logic");
```

Comment:
Why use this and instead of rejection service?

Reply:
Good catch.
Just saw flaw in code, that even with this ExceptionForProcessor, it will not work, since message passed in exception is not multipartMessage (as string) so that exception handler can handle this in correct way.
On the other side - not 100% sure what to do in case when uploading rule to UC DataApp fails - what is the correct behavior.